File: backend/app/api/sites.py
"""Site API endpoints."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from app.database.session import get_db
from app.models.models import Site, Recording, SiteType
from app.schemas.schemas import SiteCreate, SiteUpdate, SiteResponse

router = APIRouter(tags=["sites"])


@router.get("/api/projects/{project_id}/sites", response_model=list[SiteResponse])
async def list_sites(project_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Site).where(Site.project_id == project_id).order_by(Site.name))
    return result.scalars().all()


@router.post("/api/projects/{project_id}/sites", response_model=SiteResponse, status_code=201)
async def create_site(project_id: str, site: SiteCreate, db: AsyncSession = Depends(get_db)):
    db_site = Site(**site.model_dump(), project_id=project_id)
    db.add(db_site)
    await db.commit()
    await db.refresh(db_site)
    return db_site


@router.patch("/api/sites/{site_id}", response_model=SiteResponse)
async def update_site(site_id: str, updates: SiteUpdate, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Site).where(Site.id == site_id))
    site = result.scalar_one_or_none()
    if not site:
        raise HTTPException(status_code=404, detail="Site not found")
    for key, value in updates.model_dump(exclude_unset=True).items():
        setattr(site, key, value)
    await db.commit()
    await db.refresh(site)
    return site


@router.delete("/api/sites/{site_id}", status_code=204)
async def delete_site(site_id: str, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Site).where(Site.id == site_id))
    site = result.scalar_one_or_none()
    if not site:
        raise HTTPException(status_code=404, detail="Site not found")
    await db.delete(site)
    await db.commit()

File: src/api/sites.ts
/** Site API functions */

import { apiRequest } from './client'

export interface Site {
  id: string
  project_id: string
  name: string
  site_type: 'restored' | 'degraded' | 'healthy'
  latitude: number | null
  longitude: number | null
  habitat_description: string
  recorder_id: string
  recorder_model: string
  deployment_date: string | null
  mic_height_depth: number | null
  distance_to_noise: number | null
  notes: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface SiteCreate {
  name: string
  site_type: 'restored' | 'degraded' | 'healthy'
  latitude?: number
  longitude?: number
  habitat_description?: string
  recorder_id?: string
  recorder_model?: string
  deployment_date?: string
  mic_height_depth?: number
  distance_to_noise?: number
  notes?: string
  is_active?: boolean
}

export interface SiteUpdate extends Partial<SiteCreate> {}

export async function listSites(projectId: string): Promise<Site[]> {
  return apiRequest<Site[]>(`/api/projects/${projectId}/sites`)
}

export async function createSite(projectId: string, data: SiteCreate): Promise<Site> {
  return apiRequest<Site>(`/api/projects/${projectId}/sites`, { method: 'POST', body: data })
}

export async function updateSite(siteId: string, data: SiteUpdate): Promise<Site> {
  return apiRequest<Site>(`/api/sites/${siteId}`, { method: 'PATCH', body: data })
}

export async function deleteSite(siteId: string): Promise<void> {
  return apiRequest<void>(`/api/sites/${siteId}`, { method: 'DELETE' })
}
